src/main/kotlin/ai/whylabs/services/whylogs/kafka/KotlinConsumer.kt

```diff
@@ -0,0 +1,122 @@
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import org.apache.kafka.clients.consumer.ConsumerConfig
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.slf4j.LoggerFactory
+import java.time.Duration
+import kotlin.coroutines.CoroutineContext
+
+private val log = LoggerFactory.getLogger(KotlinConsumer::class.java)
+
+private fun <K, V> getConsumer(config: KotlinConsumerConfig<K, V>): KafkaConsumer<String, String> {
+    val consumerConfig = mapOf(
+        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG to config.bootstrapServers,
+        ConsumerConfig.GROUP_ID_CONFIG to config.groupId,
+        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringDeserializer",
+        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringDeserializer",
+    )
+
+    log.info("Setting up kafka consumer with config $consumerConfig")
+    val consumer = KafkaConsumer<String, String>(consumerConfig)
+    consumer.subscribe(config.topics)
+    return consumer
+}
+
+data class KotlinConsumerConfig<K, V>(
+    val consumerCount: Int = 1,
+    val topics: List<String>,
+    val bootstrapServers: List<String>,
+    val groupId: String,
+)
+
+private val mapper = jacksonObjectMapper()
+
+data class KafkaRecord<K, V>(
+    val value: V,
+    val key: K?,
+    val rawRecord: ConsumerRecord<String, String>
+)
+
+class KotlinConsumer<K, V>(
+    private val config: KotlinConsumerConfig<K, V>,
+    private val valueType: TypeReference<V>,
+    private val keyType: TypeReference<K>,
+    private val process: suspend (record: KafkaRecord<K, V>) -> Boolean
+) :
+    CoroutineScope {
+
+    companion object {
+        inline fun <reified K, reified V> consumer(
+            config: KotlinConsumerConfig<K, V>,
+            noinline processor: suspend (record: KafkaRecord<K, V>) -> Boolean
+        ): KotlinConsumer<K, V> {
+            return KotlinConsumer(config, object : TypeReference<V>() {}, object : TypeReference<K>() {}, processor)
+        }
+    }
+
+    init {
+        Runtime.getRuntime().addShutdownHook(Thread { stop() })
+    }
+
+    private val supervisorJob = SupervisorJob()
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.IO + supervisorJob
+
+    fun stop() {
+        supervisorJob.cancel()
+    }
+
+    fun start() = launch {
+        log.info("Creating ${config.consumerCount} consumers with dedicated threads.")
+        repeat(config.consumerCount) {
+            launchWorker()
+        }
+    }
+
+    // TODO need to make it reconnect if the cluster goes down
+    private fun CoroutineScope.launchWorker() = launch {
+        val consumer = getConsumer<K, V>(config)
+
+        repeatUntilCancelled(log) {
+            log.debug("Polling for messages")
+            val records = consumer.poll(Duration.ofSeconds(10))

```

Comment:
why 10 secs? wouldn't it be best to store it as a variable and change it as needed?

Reply:
It's totally arbitrary, I couldn't find a reason to prefer one time over another and I've been avoiding exposing everything possible option until someone asks for it. Less to maintain and document that way.